mono/utils/mono-threads-coop.c

```diff
@@ -67,10 +77,36 @@ mono_threads_state_poll (void)
 	}
 }
 
+static void *
+return_stack_ptr ()
+{
+	int i;
+	return &i;
+}
+
+static void
+copy_stack_data (MonoThreadInfo *info, void* stackdata_begin, void* stackdata_end)
+{
+	MonoThreadUnwindState *state;
+	int stackdata_size;
+
+	state = &info->thread_saved_state [SELF_SUSPEND_STATE_INDEX];
+
+	stackdata_size = (char*)stackdata_begin - (char*)stackdata_end;
+
+	state->gc_stackdata = mono_mempool_alloc (info->stackdata_mp, stackdata_size);
```

Comment:
Because it's initial allocation is huge and using it here out of simple convence makes little sense.

It sends the wrong message to whoever is reading this. The constant calls to mono_mempool_empty makes this a big hack.

If you feel like using a high level data structure here. There's GByteArray, which maps clearly to what you're trying to do.